diff -ruN ./pkgconfig/libjxl_cms.pc ./pkgconfig-patched/libjxl_cms.pc
--- ./pkgconfig/libjxl_cms.pc	2026-02-06 19:21:58.000000000 +0300
+++ ./pkgconfig-patched/libjxl_cms.pc	2026-02-06 23:53:27.500616382 +0300
@@ -7,7 +7,7 @@
 Description: CMS support library for libjxl
 Version: 0.11.1
 Requires: 
-Libs: -L${libdir} -ljxl_cms 
-Libs.private: -lm -lstdc++
-Cflags: -I${includedir}
-Cflags.private: -DJXL_CMS_STATIC_DEFINE
+Libs: -L${libdir} -ljxl_cms -lm -lstdc++
+Libs.private:
+Cflags: -I${includedir} -DJXL_CMS_STATIC_DEFINE
+Cflags.private:
diff -ruN ./pkgconfig/libjxl.pc ./pkgconfig-patched/libjxl.pc
--- ./pkgconfig/libjxl.pc	2026-02-06 19:21:58.000000000 +0300
+++ ./pkgconfig-patched/libjxl.pc	2026-02-06 23:56:19.676488514 +0300
@@ -6,8 +6,8 @@
 Name: libjxl
 Description: Loads and saves JPEG XL files
 Version: 0.11.1
-Requires: libhwy libbrotlienc libbrotlidec libjxl_cms
-Libs: -L${libdir} -ljxl -lm -lstdc++
+Requires: libjxl_cms
+Libs: -L${libdir} -ljxl -lhwy -lbrotlidec -lbrotlienc -lbrotlicommon -lm -lstdc++
 Libs.private: 
-Cflags: -I${includedir}
-Cflags.private: -DJXL_STATIC_DEFINE
+Cflags: -I${includedir} -DJXL_STATIC_DEFINE
+Cflags.private:
diff -ruN ./pkgconfig/libjxl_threads.pc ./pkgconfig-patched/libjxl_threads.pc
--- ./pkgconfig/libjxl_threads.pc	2026-02-06 19:21:58.000000000 +0300
+++ ./pkgconfig-patched/libjxl_threads.pc	2026-02-06 23:53:31.934664583 +0300
@@ -7,7 +7,7 @@
 Description: JPEG XL multi-thread runner using std::threads.
 Version: 0.11.1
 Requires: 
-Libs: -L${libdir} -ljxl_threads
-Libs.private: -lm
-Cflags: -I${includedir}
-Cflags.private: -DJXL_THREADS_STATIC_DEFINE
+Libs: -L${libdir} -ljxl_threads -lm -lstdc++
+Libs.private:
+Cflags: -I${includedir} -DJXL_THREADS_STATIC_DEFINE
+Cflags.private:
